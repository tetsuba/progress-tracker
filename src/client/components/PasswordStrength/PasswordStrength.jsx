/*
 * https://react-bootstrap.github.io/components/progress/
 *  */

import React from 'react';
import zxcvbn from 'zxcvbn';
import {ProgressBar} from "react-bootstrap";

const PasswordStrength = ({password}) => {
    const passwordResult = zxcvbn(password);

    const result = [
        {
            variant: 'danger',
            now: password.length > 4 ? 20: 0,
            label: password.length > 4 ? 'Bad': '',
        },
        {
            variant: 'warning',
            now: 40,
            label: 'Weak'
        },
        {
            variant: 'info',
            now: 60,
            label: 'Good'
        },
        {
            variant: 'info',
            now: 100,
            label: 'Strong'
        },
    ];

    return (
        <div>
            <ProgressBar { ...result[passwordResult.score] } />
            <small id="passwordHelpBlock" className="form-text text-muted">
                {
                    passwordResult.feedback.warning &&
                        `Tip: ${passwordResult.feedback.warning}`
                }
            </small>
        </div>
    )
};

export default PasswordStrength;