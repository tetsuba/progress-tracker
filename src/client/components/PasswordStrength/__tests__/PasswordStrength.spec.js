import { testRenderer } from '../../../../test/testHelper'
import PasswordStrength from '../PasswordStrength'

describe('<PasswordStrength>', () => {
  it('should render password strength "Bad" and variant is "danger"', () => {
    const props = {
      password: '1234',
    }
    const wrapper = testRenderer(PasswordStrength, props)
    expect(wrapper).toMatchSnapshot()
  })

  it('should render password strength "Weak" and variant is "warning"', () => {
    const props = {
      password: '1234asdf',
    }
    const wrapper = testRenderer(PasswordStrength, props)
    expect(wrapper).toMatchSnapshot()
  })

  it('should render password strength "Good" and variant is "info"', () => {
    const props = {
      password: 'J@hn2102',
    }
    const wrapper = testRenderer(PasswordStrength, props)
    expect(wrapper).toMatchSnapshot()
  })

  it('should render password strength "Strong" and variant is "info"', () => {
    const props = {
      password: '1hG6$d$Rdf',
    }
    const wrapper = testRenderer(PasswordStrength, props)
    expect(wrapper).toMatchSnapshot()
  })
})
