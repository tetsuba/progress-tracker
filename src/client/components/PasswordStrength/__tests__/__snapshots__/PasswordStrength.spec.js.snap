// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PasswordStrength> should render password strength "Bad" and variant is "danger" 1`] = `
<div>
  <ProgressBar
    animated={false}
    isChild={false}
    label=""
    max={100}
    min={0}
    now={0}
    srOnly={false}
    striped={false}
    variant="danger"
  />
  <small
    className="form-text text-muted"
    id="passwordHelpBlock"
  >
    Tip: This is a top-10 common password
  </small>
</div>
`;

exports[`<PasswordStrength> should render password strength "Good" and variant is "info" 1`] = `
<div>
  <ProgressBar
    animated={false}
    isChild={false}
    label="Good"
    max={100}
    min={0}
    now={60}
    srOnly={false}
    striped={false}
    variant="info"
  />
  <small
    className="form-text text-muted"
    id="passwordHelpBlock"
  />
</div>
`;

exports[`<PasswordStrength> should render password strength "Strong" and variant is "info" 1`] = `
<div>
  <ProgressBar
    animated={false}
    isChild={false}
    label="Strong"
    max={100}
    min={0}
    now={100}
    srOnly={false}
    striped={false}
    variant="info"
  />
  <small
    className="form-text text-muted"
    id="passwordHelpBlock"
  />
</div>
`;

exports[`<PasswordStrength> should render password strength "Weak" and variant is "warning" 1`] = `
<div>
  <ProgressBar
    animated={false}
    isChild={false}
    label="Weak"
    max={100}
    min={0}
    now={40}
    srOnly={false}
    striped={false}
    variant="warning"
  />
  <small
    className="form-text text-muted"
    id="passwordHelpBlock"
  >
    Tip: This is a very common password
  </small>
</div>
`;
